### PR TITLE
Fork choice refactor including v0.9.1. updates

### LINKED ISSUE #1261: spec v0.9.1 updates
### What is wrong?

https://github.com/ethereum/eth2.0-specs/releases/tag/v0.9.1

### How can it be fixed

Mainly removing custody data from phase 0.

Some fork choice updates.

Also some validator updates but I don't think we have that code yet

### PR DESCRIPTION
**Pull request by ralexstokes**
### What was wrong?

Closes #1261. (pretty sure)

We need to update the context for a given fork choice based on some external events. It is not straightforward to add handlers for these events so a refactor is in order.

edit: it was most natural to just go ahead and include the v0.9.1 updates in this refactor so they are in here as well.

edit: also copying the commit message from the commit in this PR as it has helpful rationale:

```
The fork choice on the beacon chain has a very dynamic quality to it
such that it makes sense to implement a "rolling" context per `StateMachine`
to track the current fork choice metadata.

1. This PR introduces the infrastructure needed to update/maintain this context.
2. Adds this context for the phase 0 LMD GHOST fork choice.
3. Implements two fork choice updates from the spec v0.9.1 release.
```

### How was it fixed?

Expose the handlers on the `Chain` abstraction and wire the event through to the underlying `StateMachine` (where a given instance of a fork choice context should live).

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: eth2/beacon/attestation_helpers.py [hunk 1]
```diff
@@ -54,14 +55,15 @@ def validate_indexed_attestation(
...
 
-    try:
-        validate_indexed_attestation_aggregate_signature(
-            state, indexed_attestation, slots_per_epoch
-        )
-    except SignatureError as error:
-        raise ValidationError(
-            f"Incorrect aggregate signature on the {indexed_attestation}", error
-        )
+    if validate_signature:
```
Comment (ralexstokes) [inline eth2/beacon/attestation_helpers.py]: this option is used downstream in LMD GHOST testing so we do not have to bother w/ constructing a valid signature on `Attestation`s going towards the fork choice test. we could remove this in a later PR if we make nicer tools for `Attestation` construction

File: eth2/beacon/chains/base.py [hunk 2]
```diff
@@ -282,6 +294,19 @@ def get_state_machine(self, at_slot: Slot = None) -> "BaseBeaconStateMachine":
     def get_genesis_state_machine_class(cls) -> Type["BaseBeaconStateMachine"]:
         return cls.sm_configuration[0][1]
 
+    # TODO how to handle the current slot in fork choice handlers
```
Comment (ralexstokes) [inline eth2/beacon/chains/base.py]: i think we will basically have a `Slot` counter in the `Node` abstraction; this `TODO` is mainly to track the idea that we will need to be very careful tracking time data given the synchrony assumptions on the network

File: eth2/beacon/db/chain.py [hunk 3]
```diff
@@ -154,6 +155,17 @@ def get_attestation_key_by_root(
...
+    # Fork choice API
+    #
+    @abstractmethod
+    def get_fork_choice_context_data_for(self, fork: str) -> bytes:
+        pass
+
+    @abstractmethod
+    def persist_fork_choice_context(self, serialized_context: bytes, fork: str) -> None:
+        pass
+
```
Comment (ralexstokes) [inline eth2/beacon/db/chain.py]: these are basically stubbed out -- serialization will come in a future PR

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 4]
```diff
@@ -1,220 +1,379 @@
...
+        raise NotImplementedError()
 
-def _take_latest_attestation_by_slot(
-    candidates: Sequence[Tuple[Slot, AttestationData]]
-) -> Tuple[Slot, AttestationData]:
-    return max(candidates, key=first)
+    @classmethod
+    def from_bytes(cls, data: bytes) -> "Context":
+        raise NotImplementedError()
+
```
Comment (ralexstokes) [inline eth2/beacon/fork_choice/lmd_ghost.py]: future PR

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 5]
```diff
@@ -1,220 +1,379 @@
...
+        target_state = self._context.checkpoint_states[target]
+
+        # NOTE: this invariant should hold based on how we handle
+        # attestation importing in the chain but we will sanity check for now
+        assert (
+            self._context.time
+            >= (attestation.data.slot + 1) * self._config.SECONDS_PER_SLOT
+        )
+
+        # TODO: has this validation already been performed?
```
Comment (ChihChengLiang) [inline eth2/beacon/fork_choice/lmd_ghost.py]: Yes, we can save this check here. We run on_attestation "upon receiving a new ``attestation`` from either within a block or directly on the wire." When received from a block, we run `validate_indexed_attestation` in `process_attestations`, and when from wire, we run it in `topic_validator.py`.

File: eth2/beacon/db/chain.py [hunk 6]
```diff
@@ -154,6 +155,17 @@ def get_attestation_key_by_root(
     def attestation_exists(self, attestation_root: HashTreeRoot) -> bool:
         pass
 
+    #
+    # Fork choice API
+    #
+    @abstractmethod
+    def get_fork_choice_context_data_for(self, fork: str) -> bytes:
+        pass
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/db/chain.py]: i was being consistent here... i'll update these in a separate PR

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 7]
```diff
@@ -1,220 +1,379 @@
...
+            sum(
+                _balance_for_validator(state, i)
+                for i in active_indices
+                if (
+                    i in self._context.latest_messages
+                    and self.get_ancestor(
+                        self._latest_message_for_index(i).root,
+                        self._get_block_by_root(root).slot,
+                    )
+                    == root
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: i'm just following `black` on this one... sorry! :(

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 8]
```diff
@@ -1,220 +1,379 @@
...
+                )
+            )
+        )
 
+    def _should_update_justified_checkpoint(
+        self, new_justified_checkpoint: Checkpoint
+    ) -> bool:
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: nice!

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 9]
```diff
@@ -1,220 +1,379 @@
...
-        except KeyError:
-            # attestation made for a block we have not imported
-            return None
-        return target_block
-
-    def _get_parent_block(self, block: BaseBeaconBlock) -> BaseBeaconBlock:
-        return self._db.get_block_by_root(block.parent_root, self._block_class)
-
-    def get_ancestor(self, block: BaseBeaconBlock, slot: Slot) -> BaseBeaconBlock:
+    def get_ancestor(self, root: SigningRoot, slot: Slot) -> Optional[Hash32]:
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: works for me

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 10]
```diff
@@ -1,220 +1,379 @@
...
+
+        self._context.blocks[root] = block
 
+        finalized_slot = self._context.finalized_slot
+        finalized_ancestor = self.get_ancestor(root, finalized_slot)
+        is_ancestor_of_finalized_block = (
+            finalized_ancestor == self._context.finalized_checkpoint.root
+        )
+        if not is_ancestor_of_finalized_block:
+            raise ValidationError(
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: the `assert`s are things that should always hold and are strict errors (most likely on our part) if they don't hold.... we should almost be able to take them out of the code and see no difference

the `ValidationError`s are things that could happen in response to events or input to our program and are times when the user should be alerted, probably to take some further action....

this being said, i'm open to unifying on `ValidationError` if this is seems more natural

File: eth2/beacon/fork_choice/lmd_ghost.py (hunk 5 above)
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: great thank you! i'll just move this to an issue for now...

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 11]
```diff
@@ -1,220 +1,379 @@
...
+        using the lexicographic ordering of the block root to break ties.
+        """
+        root = block.signing_root
 
-    block_root_score = score_block_by_root(block)
+        attestation_score = self.get_latest_attesting_balance(root)
+        block_root_score = score_block_by_root(root)
 
-    return attestation_score + block_root_score
+        return attestation_score + block_root_score
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: yeah we discussed some of this here: https://github.com/ethereum/trinity/issues/782

i'm thinking we should go ahead w/ the `Score` type and i was putting it off until i had greater reason to address it but given this PR is a non-trivial refactor -- may as well address it now :)

[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: i've started this and i think the right solution is going to add to the length of this PR...

i'll make a PR w/ these changes today but leaning towards going ahead and merging this for now

**Pull request merged**